src/main/java/ui/components/pickers/MilestonePickerState.java

```diff
@@ -30,82 +34,244 @@ public MilestonePickerState(List<PickerMilestone> milestones, String userInput)
     }
 
     /**
-     * Toggles the milestone that is the first match to the given query
-     * <p>
-     * This method is NOT case-sensitive
+     * Toggles the selection status of the first milestone in the list
      *
-     * @param milestoneList
+     * @param milestones
      */
-    private final void toggleFirstMatchingMilestone(List<PickerMilestone> milestoneList) {
-        Optional<PickerMilestone> firstMatchingMilestone = getFirstMatchingMilestone(milestoneList);
-        firstMatchingMilestone.ifPresent(this::toggleMilestone);
+    private final void toggleFirstMilestone(List<PickerMilestone> milestones) {
+        Optional<PickerMilestone> firstMilestone = milestones.stream()
+                                                            .findFirst();
+        firstMilestone.ifPresent(this::toggleMilestone);
     }
 
     /**
-     * Changes the selection statuses of all milestones in the milestone list,
-     * such that the new given milestone is toggled, and the rest are not selected
+     * Changes the selection statuses of all milestones in allMilestones,
+     * such that the given milestone is toggled, and the rest are not selected
      *
      * @param milestone
      */
     private void toggleMilestone(PickerMilestone milestone) {
-        this.currentMilestonesList.stream()
-                .forEach(listMilestone ->
-                        listMilestone.setSelected(listMilestone.equals(milestone) && !listMilestone.isSelected()));
+        allMilestones.stream()
+                .forEach(listMilestone -> listMilestone.setSelected(listMilestone.equals(milestone)
+                        && !listMilestone.isSelected()));
+    }
+
+    public List<PickerMilestone> getAllMilestones() {
+        return allMilestones;
+    }
+
+    public List<PickerMilestone> getBestMatchingMilestones() {
+        return bestMatchingMilestones;
+    }
+
+    /**
+     * Populates bestMatchingMilestones with milestones from allMilestones that best match
+     * {@code userInput}.
+     *
+     * The added milestones are references to the actual elements in allMilestones, and thus should not be
+     * unnecessarily mutated
+     * @param userInput
+     */
+    private void populateBestMatchingMilestones(String userInput) {
+        bestMatchingMilestones.clear();
+        addMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addPartiallyMatchingMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), userInput,
+                                            allMilestones);
+        }
+        if (isBelowLimit(bestMatchingMilestones, BEST_MATCHING_LIMIT)) {
+            addLikelyUnmatchedMilestones(bestMatchingMilestones, getRemainingBestMatchesToLimit(), allMilestones);
+        }
+    }
+
+    private boolean isBelowLimit(List<PickerMilestone> milestones, int limit) {
```

Comment:
Rename to `isMilestonesSizeBelowLimit` or something similar. Can't tell what limit this method checks just from current signature.